### PR TITLE
CRM-14098 Membership reminders: don't exclude people who inherit other memberships

### PR DESCRIPTION
**Pull request by agh1**
Overview
----------------------------------------
Scheduled reminders based upon memberships suppress sending to people with inherited memberships.  However, the way this was done would exclude anyone who inherited *any* membership, even if the scheduled reminder is based upon a membership that they have directly.  The only exception is if they have permission granted through the relationship.

As an example, a site may have two membership organizations and thus allow people to have two memberships in common.  One might be inherited by employees, while the other is direct memberships.

Before
----------------------------------------
If a contact inherits any membership, they will never get a scheduled reminder based upon a membership unless they have edit permissions on the relationship by which they inherit it.  This is even if it's based upon a membership that is not inherited.

After
----------------------------------------
A scheduled reminder is only excluded if

1. The membership triggering the reminder is inherited, and
2. The contact lacks relationship-based permission to edit the membership owner contact

Technical Details
----------------------------------------
The old way also generated a mile-long list of all the contact IDs that inherit any membership and put them in a WHERE clause as `AND NOT IN (...)`.  The new way makes joins based upon the actual memberships involved.

---

 * [CRM-14098: Scheduled reminders silently skipped for contacts with a non-permissioned relationship associated with ANY inherited membership type](https://issues.civicrm.org/jira/browse/CRM-14098)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I think this is the original issue https://issues.civicrm.org/jira/browse/CRM-14098

(you are doing a test? - we should be able to get @mattwire to review this PR once done)

**Comment by eileenmcnaughton** [comment]
@agh1 good digging. I'm wondering if that can somehow be turned into a code comment so future devs see that info when touching the code? (assuming you are not going to take up those points)

**Comment by agh1** [comment]
Good suggestion, @eileenmcnaughton&emdash;I just added some comments there.

**Comment by eileenmcnaughton** [comment]
@agh1 test fail relates

**Comment by agh1** [comment]
Yeah--the failure is in my new test.  However, it's passing on local, both when in isolation and with running all tests.  This may take some time to figure out.

File: Civi/ActionSchedule/RecipientBuilder.php [hunk 1]
```diff
@@ -314,7 +314,7 @@ protected function buildAddlRepeatPass() {
 
     $addlCheck = \CRM_Utils_SQL_Select::from($query['casAddlCheckFrom'])
       ->select('*')
-      ->merge($query, array('params', 'wheres'))// why only where? why not the joins?
+      ->merge($query, array('params', 'wheres', 'joins'))
```
Comment (agh1) [inline Civi/ActionSchedule/RecipientBuilder.php]: Yes, why not the joins?

Comment (eileenmcnaughton) [reply Civi/ActionSchedule/RecipientBuilder.php]: lol whose prescient code comment was that? I thought @totten when I read it but he would have said 'Dear god why...'

Comment (agh1) [reply Civi/ActionSchedule/RecipientBuilder.php]: From digging through the history, I think it was @totten's comment to himself as he rewrote the code to match the old behavior, not being certain why.

**Comment by eileenmcnaughton** [comment]
I'm at the sprint and Marcus has raised https://lab.civicrm.org/dev/core/issues/424 with me - which is a serious performance issue experienced on a small site (10k members) but with 8k of them having inherited memberships. The join that you remove here is an unindexed join and fails badly when there are lots of contacts either side of the join.

So far looking at this PR it seems good. I am getting a test fail locally on your new (yay) unit test so digging into why that is an issue

File: Civi/ActionSchedule/RecipientBuilder.php (hunk 1 above)
Comment (eileenmcnaughton) [reply Civi/ActionSchedule/RecipientBuilder.php]: I checked with Tim & yeah he was preserving existing code that didn't make sense to him :-)

**Comment by eileenmcnaughton** [comment]
@agh1 did you see my comments above - small change needed fromyou & this is good to go

**Comment by agh1** [comment]
@eileenmcnaughton here goes with the test

**Comment by agh1** [comment]
Yeah no problem--those are the old commits, just rebased onto 5.7.  I want to squash, but I just wanted to see it pass first.  Since it has always passed locally, I still have no expectation that it'll pass on the server.

**Comment by eileenmcnaughton** [comment]
@agh1 pretty green tick

**Comment by eileenmcnaughton** [comment]
@agh1 I did a squash & merge - it's not optimal from a commit history POV but I had come back to look at this again & didn't want it to drag on any more

**Comment by agh1** [comment]
Thanks @eileenmcnaughton so sorry for letting this drag.

**Comment by eileenmcnaughton** [comment]
@agh1 it's exciting that it fixes so much

**Comment by marcusjwilson** [comment]
Thanks @agh1 - Thanks @eileenmcnaughton - This should make a huge difference to one of our clients!

**Comment by eileenmcnaughton** [comment]
I hope so!

**Pull request merged**